### PR TITLE
182913315 - USER SHOULD BE ABLE TO HIGHLIGHT ON ANY TEXT IN AN ARTICLE AND MAKE COMMENTS

### PR DESCRIPTION
**Pull request by dalienst**
#### What does this Pull Request do?

- enable user to highlight text and make comments on it

#### Description of Task to be completed?

- POST: api/articles/highlight/
- GET: api/articles/highlight/
- PATCH: api/articles/highlight/
- DELETE: api/articles/highlight/<highlight_id>/

#### How should this be manually tested?

- Using postman, login user and test every endpoint stated above

#### Any background context you want to provide?

- Ability of user to highlight on a text and comment on it.
- Ability of user to reference their highlight when they need to

#### What are the relevant pivotal tracker?

- [182913315](https://www.pivotaltracker.com/n/projects/2570600/stories/182913315)

#### Screenshots (if appropriate)
Screenshot from 2022-08-08 15-39-12

Screenshot from 2022-08-08 15-39-27

Screenshot from 2022-08-08 15-47-31

#### Questions:

- N/A

### CODE REVIEW

**Review by ChrisOnsando — COMMENTED** [review]
Changes made on the PR

File: app/articles/models.py [hunk 1]
```diff
@@ -92,6 +92,29 @@ class Meta:
         ordering = ["created_at"]
 
 
+class ArticleHighlight(TimeStampedModel):
+    """
+    Highlights for any text
+    """
+
+    highlight_id = models.UUIDField(
```
Comment (mainamuriuki274) [inline app/articles/models.py:100]: Create abstract class

File: app/articles/models.py [hunk 2]
```diff
@@ -92,6 +92,29 @@ class Meta:
...
+        default=uuid.uuid4,
+        editable=False,
+        unique=True,
+        max_length=255,
+        primary_key=True,
+    )
+    highlight_start = models.PositiveIntegerField(null=True, blank=True)
```
Comment (mainamuriuki274) [inline app/articles/models.py:107]: remove null and blank =True

File: app/articles/models.py [hunk 3]
```diff
@@ -92,6 +92,29 @@ class Meta:
...
+    highlight_end = models.PositiveIntegerField(null=True, blank=True)
+    highlight_text = models.TextField(blank=True, null=True)
```
Comment (mainamuriuki274) [inline app/articles/models.py:109]: remove null and blank true

File: app/articles/serializers.py [hunk 4]
```diff
@@ -194,3 +198,90 @@ def create(self, validated_data: Any) -> Any:
...
+
+
+class TextHighlightSerializer(serializers.ModelSerializer):
+    """
+    Highlights model serializer
+    """
+
+    highlighter = UserSerializer(read_only=True)
+    article = serializers.SlugRelatedField(
+        queryset=Article.objects.all(), slug_field="slug"
```
Comment (mainamuriuki274) [inline app/articles/serializers.py:210]: use id instead of slug for lookup

File: app/articles/serializers.py [hunk 5]
```diff
@@ -194,3 +198,90 @@ def create(self, validated_data: Any) -> Any:
...
+    )
+    highlight_start = serializers.IntegerField(min_value=0, required=False)
```
Comment (mainamuriuki274) [inline app/articles/serializers.py:212]: it needs required

File: app/articles/serializers.py [hunk 6]
```diff
@@ -194,3 +198,90 @@ def create(self, validated_data: Any) -> Any:
...
+    highlight_end = serializers.IntegerField(min_value=0, required=False)
```
Comment (mainamuriuki274) [inline app/articles/serializers.py:213]: it needs required

File: app/articles/serializers.py [hunk 7]
```diff
@@ -194,3 +198,90 @@ def create(self, validated_data: Any) -> Any:
...
+        if not data.get("highlight_start") and not data.get("highlight_end"):
+            return data
+
+        if data.get("highlight_start") and not data.get("highlight_end"):
+            raise serializers.ValidationError(
+                {"highlight_end": "This field is required"},
+                code="missing_field",
+            )
+
+        if not data.get("highlight_start") and data.get("highlight_end"):
```
Comment (mainamuriuki274) [inline app/articles/serializers.py:246]: remove this

File: app/articles/serializers.py [hunk 8]
```diff
@@ -194,3 +198,90 @@ def create(self, validated_data: Any) -> Any:
...
+                },
+                code="invalid_length",
+            )
+
+        return data
+
+    def create(self, validated_data: Any) -> Any:
+        article = validated_data.get("article")
+        validated_data["highlighter"] = self.context.get("request").user  # type: ignore[union-attr]
+        if validated_data.get("highlight_start") and validated_data.get(
```
Comment (mainamuriuki274) [inline app/articles/serializers.py:273]: remove this

File: app/articles/serializers.py [hunk 9]
```diff
@@ -194,3 +198,90 @@ def create(self, validated_data: Any) -> Any:
...
+            "highlight_end"
+        ):
+            start = validated_data.get("highlight_start")
+            end = validated_data.get("highlight_end")
+            if start and end:
```
Comment (mainamuriuki274) [inline app/articles/serializers.py:278]: if not required

File: app/articles/serializers.py [hunk 10]
```diff
@@ -194,3 +198,90 @@ def create(self, validated_data: Any) -> Any:
...
+                selected = (
+                    [int(start), int(end)]
+                    if start < end
```
Comment (mainamuriuki274) [inline app/articles/serializers.py:281]: use if statement

Comment (mainamuriuki274) [reply app/articles/serializers.py:281]: if start > end:
    start, end = end, start

File: app/articles/tests/test_views.py [hunk 11]
```diff
@@ -327,3 +327,158 @@ def test_get_rate_unauthorized(self) -> None:
...
+                "highlight_start": 1,
+                "highlight_end": 6,
+                "comment": "Great work",
+            },
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(ArticleHighlight.objects.count(), count + 1)
+
+    def test_highlight_text_in_an_article_without_specifying_limits(
```
Comment (mainamuriuki274) [inline app/articles/tests/test_views.py:382]: ??

File: app/articles/tests/test_views.py [hunk 12]
```diff
@@ -327,3 +327,158 @@ def test_get_rate_unauthorized(self) -> None:
...
+        response = self.client.post(
+            reverse("highlight"),
+            data={
+                "article": self.article.slug,
+                "highlight_start": 1,
+                "comment": "Great work",
+            },
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (mainamuriuki274) [inline app/articles/tests/test_views.py:433]: assert message

Comment (mainamuriuki274) [reply app/articles/tests/test_views.py:433]: response.json()

File: app/articles/tests/test_views.py [hunk 13]
```diff
@@ -327,3 +327,158 @@ def test_get_rate_unauthorized(self) -> None:
...
+        response = self.client.post(
+            reverse("highlight"),
+            data={
+                "article": self.article.slug,
+                "highlight_end": 7,
+                "comment": "Great work",
+            },
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (mainamuriuki274) [inline app/articles/tests/test_views.py:448]: assert message

File: app/articles/tests/test_views.py [hunk 14]
```diff
@@ -327,3 +327,158 @@ def test_get_rate_unauthorized(self) -> None:
...
+            reverse("highlight"),
+            data={
+                "article": self.article.slug,
+                "highlight_start": 1000,
+                "highlight_end": 200,
+                "comment": "Great work",
+            },
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (mainamuriuki274) [inline app/articles/tests/test_views.py:466]: assert message

File: app/articles/tests/test_views.py [hunk 15]
```diff
@@ -327,3 +327,158 @@ def test_get_rate_unauthorized(self) -> None:
...
+            reverse("highlight"),
+            data={
+                "article": self.article.slug,
+                "highlight_start": 1,
+                "highlight_end": 20000,
+                "comment": "Great work",
+            },
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (mainamuriuki274) [inline app/articles/tests/test_views.py:484]: assert message

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: app/articles/models.py (hunk 1 above)
Comment (dalienst) [reply app/articles/models.py:100]: Okay

File: app/articles/models.py (hunk 2 above)
Comment (dalienst) [reply app/articles/models.py:107]: Done

File: app/articles/models.py (hunk 3 above)
Comment (dalienst) [reply app/articles/models.py:109]: Done

File: app/articles/serializers.py (hunk 4 above)
Comment (dalienst) [reply app/articles/serializers.py:210]: Done

File: app/articles/serializers.py (hunk 5 above)
Comment (dalienst) [reply app/articles/serializers.py:212]: Done

File: app/articles/serializers.py (hunk 6 above)
Comment (dalienst) [reply app/articles/serializers.py:213]: Done

File: app/articles/serializers.py (hunk 7 above)
Comment (dalienst) [reply app/articles/serializers.py:246]: Done

File: app/articles/serializers.py (hunk 8 above)
Comment (dalienst) [reply app/articles/serializers.py:273]: Done

File: app/articles/serializers.py (hunk 10 above)
Comment (dalienst) [reply app/articles/serializers.py:281]: Done

**Review by mainamuriuki274 — APPROVED** [review]

**Pull request merged**